# Insumo-Produto
_(Fonte: FONTE, André. Input-Output de Leontief: Aplicações da álgebra linear. Lisboa: 2021.)_ 

No mundo econômico, o modelo insumo-produto é uma forma de representação quantitativa da interdependência entre os setores constituintes de uma economia. Este tipo de análise desenvolvida por Wassily Leontief tem por base álgebra linear, mais propriamente, matrizes e sistemas lineares.

## Modelagem
_(Fonte: THALENBERG, Bruna; PICCHETTI, Pedro. Modelagem matemática da economia e o modelo insumo-produto de Leontief. São Paulo: 2017.)_

Antes de modelar a economia escolhida, é necessário colocar algumas premissas e
hipóteses. Tendo em vista a abrangência de setores contemplada pelo modelo insumo-produto,
adotaremos as seguintes premissas de modo a simplificar o problema, seguindo a
recomendação de Chiang e Wainwright (1980):

(i) cada indústria produz somente uma mercadoria homogênea;

(ii) cada setor utiliza uma razão fixa de insumos para a produção de seu produto;

(iii) para todos os setores os retornos de escala são constantes, isto é, a um aumento de k vezes dos insumos está associado um aumento de k vezes no produto.

A despeito do irrealismo das suposições, o modelo não perde capacidade analítica, uma vez que, se um setor produz duas mercadorias diversas ou utiliza-se de combinações de fatores diferentes, ele pode ser considerado como duas indústrias.

Temos, então, em um caso geral, que o valor do insumo requerido pelos setores
produtores para fabricar o equivalente a uma unidade monetária em mercadorias está de acordo com a tabela 1.

Tabela 1 - Representção dos valores em unidade monetária necessários de insumo para produzir uma unidade monetária de mercadoria.

Fornecedor | Setor 1 | Setor 2 | Setor 3 | ... | Setor N |
-- | -- | -- | -- | -- | -- |
Setor 1 | $a_{11}$ | $a_{12}$ | $a_{13}$ | ... | $a_{1n}$ |
Setor 2 | $a_{21}$ | $a_{22}$ | $a_{23}$ | ... | $a_{2n}$ |
Setor 1 | $a_{31}$ | $a_{32}$ | $a_{33}$ | ... | $a_{3n}$ |
: | : | : | : | ::: | : |
Setor N | $a_{n1}$ | $a_{n2}$ | $a_{n3}$ | ... | $a_{nn}$ |



Note que, para que a produção seja economicamente viável, a soma do custo dos
insumos não pode ser maior do que o valor de venda da mercadoria e, portanto, a soma dos elementos em uma coluna deve ser inferior a um. Supondo que o setor 1 fabrique o suficiente para atender os requisitos de insumos das n indústrias e à demanda final do setor aberto, seu nível de produção deve atender à equação:

$$
x_1 = a_{11}x_1 + a_{12}x_2 + ... + a_{1n}x_n + d_1.
$$

Onde $d_1$ denota a demanda final e os demais componentes da equação representam
a demanda por insumos dos demais setores. Generalizando, os níveis de produtos dos setores de uma economia devem atender a:

$$
x_n = a_{n1}x_1 + a_{n2}x_2 + ... + a_{nn}x_n + d_n.
$$

As equações dos n setores resultam em um sistema de equações, apresentado abaixo, que pode ser melhor representado em sua forma matricial.

$$
(1 - a_{11})x_1 -  a_{12}x_2 - ... - a_{1n}x_n = d_1\\
 - a_{21}x_1 -  (1 - a_{22})x_2 - ... - a_{2n}x_n = d_2\\
 ...\\
 - a_{n1}x_1 -  a_{n2}x_2 - ... - (1 - a_{nn}x_n) = d_n
$$

A matriz também pode ser escrita como $(I - A)x = d$, onde $x$ é o vetor de variáveis dependentes e $d$ o vetor da demanda final, e $(I - A)$ é chamada Matriz de Leontief.



## Teoremas
_(Fonte: FONTE, André. Input-Output de Leontief: Aplicações da álgebra linear. Lisboa: 2021.)_  

### Teorema 1 (economia rentável)

Seja $A$ a matriz insumo-produto de uma economia aberta. Caso a soma de cada coluna seja inferior a $1$, a matriz $𝐼−A$ é invertível, as entradas de $(𝐼−A)^{-1}$ são não negativas, e a economia é rentável.

### Teorema 2

Seja $A$ uma matriz input-output. Caso a matriz $𝐼 – A$ seja invertível, x existe e é único, sendo dado por:

$$
x = ( I - A )^{-1} . d
$$

## Problemas

### Problema 1 
_(Fonte: FONTE, André. Input-Output de Leontief: Aplicações da álgebra linear. Lisboa: 2021.)_ 

Considere-se o seguinte quadro económico: 3 setores industriais interdependentes ocu-pados da produção de um único tipo de produto. Sejam eles o produto A, o produto B e o pro-duto C, produzidos pelas indústrias A, B e C, respetivamente. Para a produção de cada produto, a respetiva indústria necessita de gastar não só uma certa quantidade dos outros dois produtos, como também parte da sua própria produção. Assim, suponha-se o seguinte cenário:

|  | |Insumo | | | 
| -- | -- | -- | -- | -- | 
| Produto |  | A | B | C |
| A |  | 0.10 | 0.10 | 0.40 |
| B |  | 0.40 | 0.20 | 0.10 |
| C |  | 0.20 | 0.40 | 0.05 |

Estas indústrias estão inseridas num cenário económico aberto, ou seja, estão sujeitas às exigências do mercado e a sua produção é dependente destas. Considere-se que o mercado pretende ser fornecido com  $400$ unidades monetárias do produto $A$, $200$ do produto $B$ e $100$ do produto $C$. Esses valores são representados pelo vetor $𝑑$ (do inglês demand):

$$
d =
\begin{bmatrix}
 400 \\ 200 \\ 100 
\end{bmatrix}
$$




In [2]:
# importar bibiotecas necessárias
import numpy as np
from scipy import linalg

# definir matriz insumo-produto
a = np.array([[0.1,0.1,0.4],[0.4,0.2,0.1],[0.2,0.4,0.05]])

# definir vetor de demanda
d = np.array([400, 200, 100])

# determinar o vetor de variáveis dependentes
x = np.linalg.solve(np.identity(3)-a,d)

# imprimir resutados
print(f"A quantidade de unidade monetária que cada setor deve produzir para atender a demanda de maneira mais eficiente é:\n A:{x[0]:.2f}\n B:{x[1]:.2f}\n C:{x[2]:.2f}")


A quantidade de unidade monetária que cada setor deve produzir para atender a demanda de maneira mais eficiente é:
 A:775.00
 B:708.33
 C:566.67


#### Conclusão

De acordo com a matriz de Leontieff e o algoritmo impementado, determinou-se que para que cada setor supra a demanda dos demais da forma mais efetiva o setor A deve produzir \\$775.00   o setor B deve produzir  \\$708.33  e o setor C deverá produzir \\$566.67. 

### Problema 2

_(Fonte: THALENBERG, Bruna; PICCHETTI, Pedro. Modelagem matemática da economia e o modelo insumo-produto de Leontief. São Paulo: 2017.)_

Assumiremos uma economia fictícia de seis setores, os quais apresentamos na tabela 2, e seus coeificientes são tais como descritos na tabela 3, e a demanda do setor externo, em milhões de dólares, é tal como descrita na tabela 4.

Tabela 2 - SETORES DA ECONOMIA FICTÍCIA

| Abreviação | Setor | Exemplos de produtos |
| -- | -- | -- |
| NF | Não-metal final  | Couro, móveis, alimentação |
| MF | Metal final | Maquinarias e eletrodomésticos|
| MB | Metal básico | Mineração, peças de oficina |
| NB | Não-metal básico | Vidro, madeira, produtos agrícolas |
| E | Energia | Carvão, petróleo, eletricidade |
| S | Serviços | Serviços governamentais, transporte

Tabela 3 - MATRIZ DE COEFICIENTES

|  | NF | MF | MB | NB | E | S |
| -- | -- | -- | -- | -- | -- | -- |
| NF | 0.17 | 0.004 | 0 | 0.029 | 0 | 0.008 |
| MF | 0.003 | 0.295 | 0.018 | 0.002 | 0.004 | 0.016 |
| MB | 0.025 | 0.173 | 0.46 | 0.007 | 0.011 | 0.007 
| NB | 0.346 | 0.037 | 0.021 | 0.403 | 0.011 | 0.048 |
| E | 0.007 | 0.001 | 0.039 | 0.025 | 0.358 | 0.025 |
| S | 0.12 | 0.074 | 0.104 | 0.123 | 0.173 | 0.234 |
 
Tabela 4 - DEMANDA DO SETOR EXTERNO, EM MILHÕES DE DÓLARES

|  |  |
| -- | -- |
| NF | 99640
| MF | 75548
| MB | 14444
| NB | 33501
| E | 23527
| S | 263985
 

In [4]:
# definir matriz insumo-produto
a = np.array([[0.17, 0.004, 0.0, 0.029, 0.0, 0.008], 
[0.003, 0.295, 0.018, 0.002, 0.004, 0.016],
[ 0.025, 0.173, 0.46, 0.007, 0.011, 0.007],
[0.346, 0.037, 0.021, 0.403, 0.011, 0.048],
[0.007, 0.001, 0.039, 0.025, 0.358, 0.025],
[0.12, 0.074, 0.104, 0.123, 0.173, 0.234 ]])

# definir vetor de demanda
d = np.array([99640, 75548, 14444, 33501, 23527, 263985])

# determinar o vetor de variáveis dependentes
x = np.linalg.solve(np.identity(6)-a,d)

# imprimir resutados
print(f"A quantidade de unidade monetária que cada setor deve produzir para atender a demanda de maneira mais eficiente é:\nNF:{x[0]:.2f}\n MF:{x[1]:.2f}\n MB:{x[2]:.2f}\n NB:{x[3]:.2f}\n E:{x[4]:.2f}\n S:{x[5]:.2f}\n")


A quantidade de unidade monetária que cada setor deve produzir para atender a demanda de maneira mais eficiente é:
NF:131016.45
 MF:120455.30
 MB:80671.20
 NB:178275.78
 E:66907.53
 S:431479.63



#### Conclusão

De acordo com a matriz de Leontieff e o algoritmo impementado, determinou-se que para que cada setor supra a demanda dos demais da forma mais efetiva o setor não-metal final deve produzir \\$131016.45 o setor metal final deve produzir  \\$120455.30  e o setor  matal básico deverá produzir \\$80671.20 o setor não-metal básico deve produzir \\$178275.78 o setor de energia deve produzir  \\$178275.78  e o setor de serviços deverá produzir \\$431479.63. 